src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -3078,6 +3088,303 @@ public object VisitAssignmentStatement(AssignmentStatementAst assignmentStatemen
             return Expression.Block(exprs);
         }
 
+        private Expression CompilePartialResultAssignment(AssignmentStatementAst assignmentStatementAst)
+        {
+            // The normal assignment logic won't work for us,
+            // since capture/assign presumes the entire right-hand expression must either fail or succeed.
+            // Instead, we need to create a temporary pipe,
+            // incrementally fill it with the expression result
+            // and in any event assign the result to the given variable.
+            // The logic looks like:
+            //    List<object> resultList;
+            //    Pipe oldPipe;
+            //    try {
+            //        oldPipe = funcContext._currentPipe;
+            //        resultList = new List<object>();
+            //        funcContext._currentPipe = new Pipe(resultList);
+            //        SetVariableListForTemporaryPipe(oldPipe, funcContext._currentPipe);
+            //        <rhs expression with temp pipe implicit>
+            //    } finally {
+            //        <assign lhs from result list>
+            //        funcContext._currentPipe = oldipe;
+            //    }
+
+            var resultList = NewTemp(typeof(List<object>), "resultList");
```

Comment:
FWIW if you explicitly type the variable, you can use `nameof`

```suggestion
            ParameterExpression resultList = NewTemp(typeof(List<object>), nameof(resultList));
```

Only mentioning because it drove me crazy that I couldn't use `nameof` until I found that.